front-end-tools/src/components/ReadComponent.tsx

```diff
@@ -0,0 +1,525 @@
+        setEntryPointTemplate(undefined);
+
+        let receiveTemplate;
+
+        try {
+            if (entryPointName === undefined) {
+                throw new Error('Set entry point name');
+            }
+
+            if (smartContractName === undefined) {
+                throw new Error('Set smart contract name');
+            }
+
+            let schema = '';
+
+            const schemaFromModule = deriveContractInfo ? embeddedModuleSchemaBase64 : uploadedModuleSchemaBase64;
+
+            if (schemaFromModule !== undefined) {
+                schema = schemaFromModule;
+            }
+
+            const readFunctionTemplate = getUpdateContractParameterSchema(
+                toBuffer(schema, 'base64'),
+                smartContractName,
+                entryPointName
+            );
+
+            receiveTemplate = displayTypeSchemaTemplate(readFunctionTemplate);
+
+            setEntryPointTemplate(receiveTemplate);
+        } catch (e) {
+            if (deriveContractInfo) {
+                setSchemaError(
+                    `Could not derive the embedded schema from the smart contract index. Uncheck "Derive From Smart Contract Index" checkbox to manually upload a schema or uncheck "Has Input Paramter" checkbox if this entrypoint has no input parameter. Original error: ${e}`
+                );
+            } else {
+                setSchemaError(
+                    `Could not get schema from uploaded schema. Uncheck "Has Input Paramter" checkbox if this entrypoint has no input parameter. Original error: ${e}`
+                );
+            }
+        }
+
+        if (receiveTemplate) {
+            if (inputParameterType === 'array') {
+                form.setValue('inputParameter', JSON.stringify(JSON.parse(receiveTemplate), undefined, 2));
+            } else if (inputParameterType === 'object') {
+                form.setValue('inputParameter', JSON.stringify(JSON.parse(receiveTemplate), undefined, 2));
+            }
+        }
+    }, [entryPointName, hasInputParameter, smartContractName, uploadedModuleSchemaBase64, inputParameterType]);
+
+    function onSubmit(data: FormType) {
+        setError(undefined);
+        setReturnValue(undefined);
+
+        const schema = data.deriveFromSmartContractIndex ? embeddedModuleSchemaBase64 : uploadedModuleSchemaBase64;
+
+        // Invoke smart contract (read)
+
+        const promise = read(
+            client,
+            data.smartContractName,
+            BigInt(data.smartContractIndex),
+            data.entryPointName,
+            schema,
+            data.inputParameter,
+            data.inputParameterType,
+            data.hasInputParameter,
+            data.deriveFromSmartContractIndex
+        );
+
+        promise
+            .then((value) => {
+                setReturnValue(value);
+            })
+            .catch((err: Error) => setError((err as Error).message));
+    }
+
+    return (
+        <Box header="Read From Smart Contract">
```

Comment:
I think this is not really the whole story. You can **simulate** a contract call in this flow, not just getting back return values.

I think that should be emphasised as well.

Reply:
Currently, this box is set up to actually read a value from the smart contract and display the `return_value` to the user (either the raw bytes or deserialized if a schema is available). No dry run/simulation can be done with the box.

Also, the `invoke` call can be used to dry-run an `update` transaction, such transactions usually have no `return_value`, so I wouldn't want to treat them in this box. Maybe a separate `simulation` box would be more appropriate.

If I recall correctly we intended to add `dry-run` steps before the deploy/init/update transactions at some future steps which would be the best option in my opinion.
 
 I start collection feature requests here:
 https://github.com/Concordium/concordium-smart-contract-tools/issues/131

Reply:
Added it to:
https://github.com/Concordium/concordium-smart-contract-tools/issues/131
```
Better error handling. In particular, think about using the read box as a simulation/dry-run box.
https://github.com/Concordium/concordium-smart-contract-tools/pull/123#discussion_r1387772430
```